subxt/src/storage/storage_address.rs

```diff
@@ -54,64 +58,314 @@ pub struct Yes;
+    PartialOrd(bound = "Keys: std::cmp::PartialOrd")
 )]
-pub struct Address<StorageKey, ReturnTy, Fetchable, Defaultable, Iterable> {
+pub struct Address<Keys: StorageMultiKey, ReturnTy, Fetchable, Defaultable, Iterable> {
     pallet_name: Cow<'static, str>,
     entry_name: Cow<'static, str>,
-    storage_entry_keys: Vec<StorageKey>,
+    keys: Keys,
     validation_hash: Option<[u8; 32]>,
     _marker: std::marker::PhantomData<(ReturnTy, Fetchable, Defaultable, Iterable)>,
 }
 
+/// A storage key, mostly used for static encoded values.
+/// The original value is only given during construction, but can be
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""))]
+pub struct StorageKey<K: ?Sized> {
+    bytes: Static<Encoded>,
+    _marker: std::marker::PhantomData<K>,
+}
+
+impl<K: codec::Encode + ?Sized> StorageKey<K> {
+    /// Creates a new static storage key
+    pub fn new(key: &K) -> Self {
+        StorageKey {
+            bytes: Static(Encoded(key.encode())),
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Returns the scale-encoded bytes that make up this key
+    pub fn bytes(&self) -> &[u8] {
+        &self.bytes.0 .0
+    }
+}
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageMultiKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType>;
+
+    /// Attempts to decode the StorageMultiKey from a whole storage address.
+    /// The key/keys can only be recovered if all hashers are concat-style hashers.
+    /// Example: Imagine The `StorageMultiKey` is a tuple (A,B) and the hashers are: [Blake2_128Concat, Twox64Concat].
+    /// Then the memory layout of the storage key is:
+    /// ```txt
+    /// | 8 bytes pallet hash | 8 bytes entry hash | 16 bytes hash of A | ... bytes of A | 8 bytes hash of B | ... bytes of B |
+    /// ```
+    /// Returns None, if any of the hashers is not a concat-style hasher.
+    fn decode_from_address_bytes(
+        address_bytes: &[u8],
+        hashers_and_ty_ids: &[(StorageHasher, u32)],
+        metadata: &Metadata,
+    ) -> Option<Result<Self, Error>>
+    where
+        Self: Sized + 'static;
+}
+
+/// Implement `StorageMultiKey` for `()` which can be used for keyless storage entries
+impl StorageMultiKey for () {
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType> {
+        // Note: this returns the storage root address of the storage entry.
+        // It gives the same result as if you were to use `vec![]` as a `StorageMultiKey`.
+        std::iter::empty()
+    }
+
+    fn decode_from_address_bytes(
+        address_bytes: &[u8],
+        _hashers_and_ty_ids: &[(StorageHasher, u32)],
+        _metadata: &Metadata,
+    ) -> Option<Result<Self, Error>> {
+        if address_bytes.len() < 16 {
```

Comment:
Nit: I would add a comment that for a valid storage key, we'd need at least 8 bytes for the pallet hash and 8 for the entry